packages/builtin/array.pony

```diff
@@ -328,14 +331,40 @@ class Array[A] is Seq[A]
 
         n = n + 1
       end
+    end
+
+    n = 0
+
+    // If a concrete len is specified, we take the caller at their word
```

Comment:
Actually, this is still memory safe for all inputs.  If there are fewer than `len` elements available from the iterator, then we might `reserve` too much, but we will never `reserve` too little.

The worst case scenario we can have with the caller is lying about the number of elements available is that a huge amount of space will be reserved, but never actually used.  And this is already possible already if the caller allocates an `Array` with a huge value in the constructor argument.